services/storage/eval.go

```diff
@@ -0,0 +1,257 @@
+		}
+	}
+
+	// Evaluate if both sides are simple types.
+	switch lhs := lhs.(type) {
+	case bool:
+		rhs, ok := rhs.(bool)
+		switch expr.Op {
+		case influxql.AND:
+			return ok && (lhs && rhs)
+		case influxql.OR:
+			return ok && (lhs || rhs)
+		case influxql.BITWISE_AND:
+			return ok && (lhs && rhs)
+		case influxql.BITWISE_OR:
+			return ok && (lhs || rhs)
+		case influxql.BITWISE_XOR:
+			return ok && (lhs != rhs)
+		case influxql.EQ:
+			return ok && (lhs == rhs)
+		case influxql.NEQ:
+			return ok && (lhs != rhs)
+		}
+	case float64:
+		// Try the rhs as a float64 or int64
+		rhsf, ok := rhs.(float64)
+		if !ok {
+			var rhsi int64
+			if rhsi, ok = rhs.(int64); ok {
+				rhsf = float64(rhsi)
+			}
+		}
+
+		rhs := rhsf
+		switch expr.Op {
+		case influxql.EQ:
+			return ok && (lhs == rhs)
+		case influxql.NEQ:
+			return ok && (lhs != rhs)
+		case influxql.LT:
+			return ok && (lhs < rhs)
+		case influxql.LTE:
+			return ok && (lhs <= rhs)
+		case influxql.GT:
+			return ok && (lhs > rhs)
+		case influxql.GTE:
+			return ok && (lhs >= rhs)
+		case influxql.ADD:
+			if !ok {
+				return nil
+			}
+			return lhs + rhs
+		case influxql.SUB:
+			if !ok {
+				return nil
+			}
+			return lhs - rhs
+		case influxql.MUL:
+			if !ok {
+				return nil
+			}
+			return lhs * rhs
+		case influxql.DIV:
+			if !ok {
+				return nil
+			} else if rhs == 0 {
+				return float64(0)
+			}
+			return lhs / rhs
+		case influxql.MOD:
+			if !ok {
+				return nil
+			}
+			return math.Mod(lhs, rhs)
+		}
+	case int64:
+		// Try as a float64 to see if a float cast is required.
+		rhsf, ok := rhs.(float64)
+		if ok {
+			lhs := float64(lhs)
+			rhs := rhsf
+			switch expr.Op {
+			case influxql.EQ:
+				return lhs == rhs
+			case influxql.NEQ:
+				return lhs != rhs
+			case influxql.LT:
+				return lhs < rhs
+			case influxql.LTE:
+				return lhs <= rhs
+			case influxql.GT:
+				return lhs > rhs
+			case influxql.GTE:
+				return lhs >= rhs
+			case influxql.ADD:
+				return lhs + rhs
+			case influxql.SUB:
+				return lhs - rhs
+			case influxql.MUL:
+				return lhs * rhs
+			case influxql.DIV:
+				if rhs == 0 {
+					return float64(0)
+				}
+				return lhs / rhs
+			case influxql.MOD:
+				return math.Mod(lhs, rhs)
+			}
+		} else {
+			rhs, ok := rhs.(int64)
+			switch expr.Op {
+			case influxql.EQ:
+				return ok && (lhs == rhs)
+			case influxql.NEQ:
+				return ok && (lhs != rhs)
+			case influxql.LT:
+				return ok && (lhs < rhs)
+			case influxql.LTE:
+				return ok && (lhs <= rhs)
+			case influxql.GT:
+				return ok && (lhs > rhs)
+			case influxql.GTE:
+				return ok && (lhs >= rhs)
+			case influxql.ADD:
+				if !ok {
+					return nil
+				}
+				return lhs + rhs
+			case influxql.SUB:
+				if !ok {
+					return nil
+				}
+				return lhs - rhs
+			case influxql.MUL:
+				if !ok {
+					return nil
+				}
+				return lhs * rhs
+			case influxql.DIV:
+				if !ok {
+					return nil
+				} else if rhs == 0 {
+					return float64(0)
```

Comment:
I'm not familiar with this part of the codebase—what's the reason for returning a `float64` when the result of the division is undefined?

Reply:
`eval.go` was copied from the `query` package. 

After the MVP I plan to replace eval.go with some form of expression compiler, so we're not interpreting an AST each time. It is pretty inefficient.